src/lib.rs

```diff
@@ -77,22 +80,55 @@ impl Program {
     }
 }
 
-#[derive(Clone, Debug, Default, Eq, PartialEq)]
-pub struct KnowledgeBase {
-    pub atoms: HashSet<Atom>,
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct KnowledgeBase<P, V>
+where
+    P: Clone + Eq + Hash,
+    V: Clone + Eq + Hash,
```

Comment:
These bounds can be simplified a little by implementing 
```
pub trait Predicate: Clone + Eq + Hash { }
pub trait Value: Clone + Eq + Hash {  }

impl Predicate for T where T: Clone + Eq + Hash  {  }
impl Value for T where T: Clone + Eq + Hash  {  }
```

and requiring `P: Predicate, V: Value`